scripts/build_osm_network.py

```diff
@@ -0,0 +1,1051 @@
+    if not df_transformers.empty:
+        init_index = 0 if lines.empty else lines.index[-1] + 1
+        df_transformers.set_index(init_index + df_transformers.index, inplace=True)
+    # update line endings
+    df_transformers = line_endings_to_bus_conversion(df_transformers)
+
+    return df_transformers
+
+
+def get_converters(buses):
+    """
+    Function to create fake converter lines that connect buses of the same
+    station_id of different polarities.
+    """
+
+    df_converters = []
+
+    for g_name, g_value in buses.sort_values("voltage", ascending=True).groupby(
+        by="station_id"
+    ):
+        # note: by construction there cannot be more that two buses with the same station_id and same voltage
+        n_voltages = len(g_value)
+
+        # A converter stations should have both AC and DC parts
+        if g_value["dc"].any() & ~g_value["dc"].all():
+            dc_voltage = g_value[g_value.dc]["voltage"].values
+
+            for u in dc_voltage:
+                id_0 = g_value[g_value["dc"] & g_value["voltage"].isin([u])].index[0]
+
+                ac_voltages = g_value[~g_value.dc]["voltage"]
+                # A converter is added between a DC nodes and AC one with the closest voltage
+                id_1 = ac_voltages.sub(u).abs().idxmin()
+
+                geom_conv = LineString(
+                    [g_value.geometry.loc[id_0], g_value.geometry.loc[id_1]]
+                )
+
+                # check if bus is a dclink boundary point, only then add converter
+                df_converters.append(
+                    [
+                        f"convert_{g_name}_{id_0}",  # "line_id"
+                        g_value["bus_id"].loc[id_0],  # "bus0"
+                        g_value["bus_id"].loc[id_1],  # "bus1"
+                        False,  # "underground"
+                        False,  # "under_construction"
+                        g_value.country.loc[id_0],  # "country"
+                        geom_conv,  # "geometry"
+                    ]
+                )
+
+    # name of the columns
+    conv_columns = [
+        "converter_id",
+        "bus0",
+        "bus1",
+        "underground",
+        "under_construction",
+        "country",
+        "geometry",
+    ]
+
+    df_converters = gpd.GeoDataFrame(df_converters, columns=conv_columns).reset_index()
+
+    return df_converters
+
+
+def connect_stations_same_station_id(lines, buses):
+    """
+    Function to create fake links between substations with the same
+    substation_id.
+    """
+    ac_freq = get_ac_frequency(lines)
+    station_id_list = buses.station_id.unique()
+
+    add_lines = []
+    from shapely.geometry import LineString
+
+    for s_id in station_id_list:
+        buses_station_id = buses[buses.station_id == s_id]
+
+        if len(buses_station_id) > 1:
+            for b_it in range(1, len(buses_station_id)):
+                add_lines.append(
+                    [
```

Comment:
double-check if performance bottleneck

do you see the possible to reduce the nesting levels? E.g.

```py
if len(buses_station_id) <= 1:
    continue
```

Reply:
open todo

Reply:
not necessary